### PR TITLE
Add gitlab support in scripts/new.lua

### PR DESCRIPTION
**Pull request by Tomcat-42**
- Add gitlab support
- Refactor scripts/new.lua

### CODE REVIEW

**Comment by Tomcat-42** [comment]
I gave more context in the issue #2137

File: scripts/new.lua [hunk 1]
```diff
@@ -8,160 +8,284 @@ import("devel.git")
...
+    local packagename = assert(data.name, "package name not found!"):lower()
+    local packagefile = path.join("packages", string.sub(packagename, 1, 1), packagename, "xmake.lua")
+    local file = io.open(packagefile, "w")
+
+    -- Define package and homepage
+    file:write(string.format('package("%s")\n', packagename))
+    local homepage = data.homepageUrl and data.homepageUrl ~= "" and data.homepageUrl or data.url
     if homepage then
-        file:print('    set_homepage("%s")', homepage)
+        file:write(string.format('    set_homepage("%s")\n', homepage))
```
Comment (waruqi) [inline scripts/new.lua:118]: why not use print? it will add newline.

File: scripts/new.lua [hunk 2]
```diff
@@ -8,160 +8,284 @@ import("devel.git")
...
         repodir = tmpfile .. ".dir"
-        print("downloading %s", url)
+
+        file:write(string.format('    add_urls("https://%s/%s/-/archive/$(version).tar.gz",\n', host, reponame))
+        file:write(string.format('             "%s")\n', giturl))
+
+        print(string.format("downloading %s", url))
         http.download(url, tmpfile)
-        file:print('    add_versions("%s", "%s")', latest_release.tagName, hash.sha256(tmpfile))
+
```
Comment (waruqi) [inline scripts/new.lua:158]: please use file:print

File: scripts/new.lua [hunk 3]
```diff
@@ -8,160 +8,284 @@ import("devel.git")
...
-        print("downloading %s", giturl)
-        git.clone(giturl, {outputdir = repodir, depth = 1})
-        local commit = git.lastcommit({repodir = repodir})
-        local version = try{ function() return os.iorunv("git", {"log", "-1", "--date=format:%Y.%m.%d", "--format=%ad"}, {curdir = repodir}) end}
+        local giturl = string.format("git@%s:%s.git", host, reponame)
+        repodir = os.tmpfile({ ramdisk = false })
+
+        file:write(string.format('    add_urls("%s")\n', giturl))
+
+        print(string.format("downloading %s", giturl))
```
Comment (waruqi) [inline scripts/new.lua:168]: please remove string.format, print support it.

File: scripts/new.lua [hunk 4]
```diff
@@ -8,160 +8,284 @@ import("devel.git")
...
-        repodir = os.tmpfile({ramdisk = false})
-        file:print('    add_urls("%s")', giturl)
-        print("downloading %s", giturl)
-        git.clone(giturl, {outputdir = repodir, depth = 1})
-        local commit = git.lastcommit({repodir = repodir})
-        local version = try{ function() return os.iorunv("git", {"log", "-1", "--date=format:%Y.%m.%d", "--format=%ad"}, {curdir = repodir}) end}
+        local giturl = string.format("git@%s:%s.git", host, reponame)
+        repodir = os.tmpfile({ ramdisk = false })
+
+        file:write(string.format('    add_urls("%s")\n', giturl))
```
Comment (waruqi) [inline scripts/new.lua:166]: file:print

File: scripts/new.lua [hunk 5]
```diff
@@ -8,160 +8,284 @@ import("devel.git")
...
+    {
+        nil,
+        "repo",
+        "v",
+        nil,
+        "Set repository name.",
+        "e.g. ",
+        "  - github:xmake-io/xmake",
+        "  - gitlab:xmake-io/xmake",
+    },
```
Comment (waruqi) [inline scripts/new.lua:20]: Do not change the indentation format of the code, please ensure minimal code changes.

File: scripts/new.lua [hunk 6]
```diff
@@ -8,160 +8,284 @@ import("devel.git")
...
 
-    -- generate package header
-    local packagename = assert(repoinfo.name, "package name not found!"):lower()
-    local packagefile = path.join("packages", packagename:sub(1, 1), packagename, "xmake.lua")
-    local file = io.open(packagefile, "w")
-    file:print('package("%s")', packagename)
-    local homepage = repoinfo.homepageUrl
-    if homepage == nil or homepage == "" then
-        homepage = repoinfo.url
+    -- Return host and data
```
Comment (waruqi) [inline scripts/new.lua:81]: Please do not add too many redundant comments.

**Comment by Tomcat-42** [comment]
Hello @waruqi, thanks for your feedback!!!

I fixed your considerations in my code, have a look at it

File: scripts/new.lua [hunk 7]
```diff
@@ -11,174 +11,286 @@ local options = {
...
+        local url = string.format("https://%s/%s/archive/refs/tags/%s.tar.gz", host, reponame, latest_release.tagName)
+        local giturl = string.format("https://%s/%s.git", host, reponame)
+        local tmpfile = os.tmpfile({ ramdisk = false }) .. ".tar.gz"
         repodir = tmpfile .. ".dir"
-        print("downloading %s", url)
+
+        file:print('    add_urls("https://%s/%s/-/archive/$(version).tar.gz', host, reponame)
+        file:print('             "%s")\n', giturl)
+
+        print(string.format("downloading %s", url))
```
Comment (waruqi) [inline scripts/new.lua:117]:
```
print("downloading %s", url)
```

File: scripts/new.lua [hunk 8]
```diff
@@ -11,174 +11,286 @@ local options = {
...
+        local giturl = string.format("git@%s:%s.git", host, reponame)
+        repodir = os.tmpfile({ ramdisk = false })
+
         file:print('    add_urls("%s")', giturl)
-        print("downloading %s", giturl)
-        git.clone(giturl, {outputdir = repodir, depth = 1})
-        local commit = git.lastcommit({repodir = repodir})
-        local version = try{ function() return os.iorunv("git", {"log", "-1", "--date=format:%Y.%m.%d", "--format=%ad"}, {curdir = repodir}) end}
+
+        print(string.format("downloading %s", giturl))
```
Comment (waruqi) [inline scripts/new.lua:129]: and here

File: scripts/new.lua [hunk 9]
```diff
@@ -11,174 +11,286 @@ local options = {
...
-        "Usage: xmake l scripts/new.lua [options]")
+    local opt = option.parse(table.pack(...), options, "New a package.", "", "Usage: xmake l scripts/new.lua [options]")
     local repo = opt.repo
-    if repo and repo:startswith("github:") then
-        _generate_package_from_github(repo:sub(8))
-    else
-        raise("we need set repository name first!")
+
+    if not repo then
+        error("Repository name must be set!")
```
Comment (waruqi) [inline scripts/new.lua:280]: please use raise instead of error

File: scripts/new.lua [hunk 10]
```diff
@@ -11,174 +11,286 @@ local options = {
...
+        generate_package(reponame, get_github_data)
+        return
+    end
+
+    if repo:startswith("gitlab:") then
+        generate_package(reponame, get_gitlab_data)
+        return
+    end
+
+    error("Unsupported repository source. Only 'github' and 'gitlab' are supported.")
```
Comment (waruqi) [inline scripts/new.lua:295]: and here

File: scripts/new.lua [hunk 11]
```diff
@@ -11,174 +11,286 @@ local options = {
...
-        file:print('            target("%s")', packagename)
-        file:write('                set_kind("$(kind)")\n')
-        file:print('                add_files("src/*.c")')
-        file:print('                add_headerfiles("src/(*.h)")')
-        file:print('        ]])')
-        file:print('        if package:config("shared") then')
-        file:print('            configs.kind = "shared"')
-        file:print('        end')
-        file:print('        import("package.tools.xmake").install(package, configs)')
+    -- file:print('')
```
Comment (waruqi) [inline scripts/new.lua:255]: you need enable this line

```
file:print('')
```

File: scripts/new.lua [hunk 12]
```diff
@@ -11,174 +11,286 @@ local options = {
...
-    elseif need_autogen then
-        file:print("")
-        file:print('    add_deps("autoconf", "automake", "libtool")')
-    elseif has_bazel then
-        file:print("")
-        file:print('    add_deps("bazel")')
+    if build_system then
+        file:print('')
+        for _, dep in ipairs(build_system.deps) do
+            file:print('    add_deps("' .. dep .. '")')
```
Comment (waruqi) [inline scripts/new.lua:250]: please add deps in single line. like this `add_deps("autoconf", "automake", "libtool")`

you can use `table.concat(deps, "\", ")`

**Comment by Tomcat-42** [comment]
Hello @waruqi, I was a little busy since our last review, Thanks for fixing the issues tthat you pointed.

Can we add some oother features, or do you think that we can merge and close the issue?

Looking forward for your answer.

**Comment by waruqi** [comment]
I have modified and merged it.

**Pull request merged**